app/models/pfmp.rb

```diff
@@ -47,6 +47,10 @@ def state_machine
     end
   end
 
+  def validate!
+    transition_to!(:validated)
+  end
```

Comment:
Pour la consistance je n'introduirais pas de sugar method sur la SM, c'est à mon avis plus clair d'utiliser le `transition_to!`.

Le cas contraire cognitivement avec cette gem on ne capterai pas forcément qu'il s'agit d'une transition de state machine si ces sugars methods ne sont pas généralisées.

D'autant plus qu'ici `validate!` peut faire penser à une méthode d'ActiveRecord, ce qui peut être confusant.

Reply:
deux bénéfices :

- le contrôleur demande juste de valider une PFMP, et ne sait pas (tant mieux) que ça induit un changement de l'automate (ni l'interface qui va avec, `transition_to`) ;
- avec cette approche de `setter` on va pouvoir rajouter la logique autour bientôt (fruit d'une autre PR).

Le naming n'est pas dingue je te l'accorde

Reply:
Tes 2 bénéfices le sont si on globalise cette approche de wrap de méthode. J'utilise une autre gem qui utilise cette approche (`event!`) et en pratique je trouve ça moins clair.

Par ailleurs je wrapperai plus ça dans un pattern chain of command, car en pratique effectuer des callbacks dans une machine à états c'est la même que des callbacks ActiveRecord: ça devient vite bordélique et avec des side effects non voulu dans certaines situations. Bonus point: tu peux tester facilement en isolation.

Reply:
elle l'est déjà non ? https://github.com/betagouv/aplypro/blob/payments/allowance-rework/app/models/payment.rb#L26-L40

Reply:
Partiellement en effet, faudrait la mixin avec de la metaprog (`states.each do |state| define_method("#{state}!") { transition_to(state) }`) comme ça c'est clair.

Reply:
le trigger n'a pas forcément le même nom que le state

Reply:
C'était pour l'exemple, faut en effet loop sur les events.
Ou alors changer de gem qui fait déjà ce taff (i.e. quasiment toutes les autres)

Reply:
Y'a deux débats là :
1. la sugar method `validate!`
2. le `after_save` qui fait la transition

Mon avis :
1. sugar method ou pas, les deux me vont. C'est assez peu important à mes yeux.
2. je suis d'accord avec @skelz0r sur le fait que les trucs qui créent de la data en after_save, c'est à éviter. Je donnerait la responsabilité de créer le paiement à l'objet qui déclenche le changement de state de la pfmp. Quitte à faire un `PfmpTransitionManager` ou un truc comme ça.

Reply:
c'est pas une sugar méthode 😄 c'est du SOLID en 1. ayant un seul endroit pour réunir le changement 2. cacher les détails aux potentiels callers.

Les machines à états sont spécialistes du callback et des guards, c.f les APIs de Statesman et du tout aussi excellent AASM. Ne pas mélanger le layer de persistence on est bien d'accord, c'est le danger d'en faire trop avec ActiveRecord.

Et en même ce n'est pas _du tout_ l'objet de cette PR

Reply:
Vu ensemble à l'oral.

Je suis d'accord pour dire que générer un paiement en `after_save` c'est bof, mais on va pas le traiter dans cette PR.

On va garder la méthode `validate!` aussi, c'est quand même cohérent de pouvoir dire "valide ma pfmp wesh" dans le controlleur, sans parler de transitions.